s3store/s3store.go

```diff
@@ -562,27 +561,39 @@ func isAwsError(err error, code string) bool {
 	return false
 }
 
-func (store S3Store) CalcOptimalPartSize(size *int64) int64 {
+func (store S3Store) CalcOptimalPartSize(size int64) (optimalPartSize int64, err error) {
+	var errorstrings []string
+
 	switch {
-	// We can only manage files up to MaxObjectSize, else we need to fail.
-	case *size > store.MaxObjectSize:
-		return 0
 	// When upload is smaller or equal MinPartSize, we upload in just one part.
-	case *size <= store.MinPartSize:
-		return store.MinPartSize
+	case size <= store.MinPartSize:
+		optimalPartSize = store.MinPartSize
 	// When we need 9999 parts or less with MinPartSize.
-	case *size/store.MinPartSize < store.MaxMultipartParts:
-		return store.MinPartSize
-	// When we can divide our upload into exactly MaxMultipartParts parts with
-	// no bytes leftover, we will not need an spare last part.
-	// Also, when MaxObjectSize is equal to MaxPartSize * MaxMultipartParts
+	case size/store.MinPartSize < store.MaxMultipartParts:
+		optimalPartSize = store.MinPartSize
+	// If our upload divides up exactly into MaxMultipartParts parts with
+	// no bytes leftover, we will not need an spare last part. So we can go with
+	// a straight division. Otherwise we would exceed MaxPartSize in a scenario,
+	// where MaxObjectSize is equal to MaxPartSize * MaxMultipartParts
 	// (which is not the case with the current AWS S3 API specification, but
-	// might be in the future or with other S3-aware stores), we need this in
-	// order for our Multipart-Upload to reach full MaxObjectSize.
-	case int64(math.Mod(float64(*size), float64(store.MaxMultipartParts))) == 0:
-		return *size / store.MaxMultipartParts
-	// In all other cases, we need a spare last piece for the remaining bytes.
+	// might be in the future or with other S3-aware stores).
+	case size%store.MaxMultipartParts == 0:
+		optimalPartSize = size / store.MaxMultipartParts
+	// In all other cases, we need to round up to the next integer.
 	default:
-		return *size / (store.MaxMultipartParts - 1)
+		optimalPartSize = size/store.MaxMultipartParts + 1
+	}
+
+	// an upload larger than MaxObjectSize must throw an error
+	if size > store.MaxObjectSize {
+		errorstrings = append(errorstrings, fmt.Sprintf("CalcOptimalPartSize: upload size of %v bytes exceeds MaxObjectSize of %v bytes", size, store.MaxObjectSize))
```

Comment:
I wanted optimalPartSize still to be calculated and returned in any case, as it helps to debug, how the algorithm is behaving. Especially useful for the tests.

Reply:
IMO there is no need to calculate a part size if the upload exceeds the maximum object size because the upload will never be possible. It's confusion why this check is down here.

Reply:
OK. So I return part size 0 and the error then. Coming in a second.

Reply:
I will have to restructure the tests again to accommodate that change. One more commit coming.

Reply:
I am too quick to push. One more fix pending.